sqlserver/datadog_checks/sqlserver/connection.py

```diff
@@ -103,6 +105,41 @@ def parse_connection_string_properties(cs):
     return params
 
 
+known_hresult_codes = {
+    -2147352567: "unable to connect",
+    -2147217843: "login failed for user",
+    # this error can also be caused by a failed TCP connection but we are already reporting on the TCP
+    # connection status via test_network_connectivity so we don't need to explicitly state that
+    # as an error condition in this message
+    -2147467259: "could not open database requested by login",
+}
+
+
+def _format_connection_exception(e):
+    """
+    Formats the provided database connection exception.
+    If the exception comes from an ADO Provider and contains a misleading 'Invalid connection string attribute' message
+    then the message is replaced with more descriptive messages based on the contained HResult error codes.
+    """
+    if adodbapi is not None:
```

Comment:
does the ODBC driver give useful feedback when it fails? have we tested that too?

Reply:
ah okay, good to know